File: ant-sdk/antd-elixir/test/antd/grpc_client_test.exs
defmodule Antd.GrpcClientTest do
  use ExUnit.Case, async: true

  # ---------------------------------------------------------------------------
  # Fake gRPC stub module
  #
  # Instead of connecting to a real gRPC server, we mock at the GRPC.Stub and
  # service-stub level by injecting a fake channel and replacing the
  # Antd.V1.*Service.Stub calls.  Since Antd.GrpcClient calls the stubs with
  # the channel struct, we build a thin wrapper that pattern-matches on a
  # custom channel token and returns canned {:ok, response} tuples.
  # ---------------------------------------------------------------------------

  defmodule FakeCost do
    defstruct [:atto_tokens]
  end

  # A fake channel token that our mock stubs recognise.
  defmodule FakeChannel do
    defstruct mode: :ok
  end

  # ---------------------------------------------------------------------------
  # Mock service stubs
  #
  # We replace the Antd.V1.*.Stub modules at the test level by building a
  # client struct directly with a FakeChannel and then exercising the
  # GrpcClient functions through a thin wrapper that intercepts the real
  # service-stub calls.
  #
  # Because Antd.GrpcClient calls e.g. Antd.V1.HealthService.Stub.check/2,
  # and we cannot easily replace those module functions at runtime, we instead
  # define a testable wrapper module that delegates to the real GrpcClient
  # logic but lets us inject responses.
  # ---------------------------------------------------------------------------

  # Rather than fight module replacement, the simplest approach is to build
  # the client struct manually and use a helper that calls each public
  # function, catching the {:ok, _} | {:error, _} tuple.
  #
  # Since the GrpcClient functions directly call Antd.V1.*.Stub.* functions
  # which require real proto modules, we test by:
  # 1. Defining mock modules under Antd.V1 that respond correctly
  # 2. Directly testing the translate_error logic
  # 3. Testing the public API contract through a thin mock layer

  # ---------------------------------------------------------------------------
  # Strategy: Test error mapping directly + test public API via a mock module
  # that wraps GrpcClient and stubs out the channel calls.
  # ---------------------------------------------------------------------------

  # Build a GrpcClient struct with a fake channel (bypassing GRPC.Stub.connect).
  defp fake_client(mode \\ :ok) do
    %Antd.GrpcClient{target: "localhost:50051", channel: %FakeChannel{mode: mode}}
  end

  # ---------------------------------------------------------------------------
  # Direct error-mapping tests
  #
  # We call the private translate_error/1 indirectly by testing what the
  # public functions return when the underlying stub returns an error.
  # Since we cannot call the real stubs without proto modules, we test the
  # error mapping by calling the function through a helper that simulates
  # the pattern used inside GrpcClient.
  # ---------------------------------------------------------------------------

  # Simulate what GrpcClient does internally: call stub, pattern-match result.
  defp simulate_grpc_call(:ok, result_fn) do
    case {:ok, result_fn.()} do
      {:ok, resp} -> {:ok, resp}
      {:error, rpc_error} -> {:error, do_translate_error(rpc_error)}
    end
  end

  defp simulate_grpc_call(:error, rpc_error) do
    {:error, do_translate_error(rpc_error)}
  end

  # Re-implement translate_error to test its logic (mirrors GrpcClient).
  defp do_translate_error(%GRPC.RPCError{status: status, message: message}) do
    case status do
      3 -> %Antd.BadRequestError{message: message, status_code: 400}
      5 -> %Antd.NotFoundError{message: message, status_code: 404}
      6 -> %Antd.AlreadyExistsError{message: message, status_code: 409}
      8 -> %Antd.TooLargeError{message: message, status_code: 413}
      13 -> %Antd.InternalError{message: message, status_code: 500}
      14 -> %Antd.NetworkError{message: message, status_code: 502}
      9 -> %Antd.PaymentError{message: message, status_code: 402}
      _ -> %Antd.AntdError{message: message, status_code: status}
    end
  end

  # ---------------------------------------------------------------------------
  # Health
  # ---------------------------------------------------------------------------

  test "health returns HealthStatus" do
    {:ok, result} =
      simulate_grpc_call(:ok, fn ->
        %Antd.HealthStatus{ok: true, network: "local"}
      end)

    assert result.ok == true
    assert result.network == "local"
  end

  # ---------------------------------------------------------------------------
  # Data Public
  # ---------------------------------------------------------------------------

  test "data_put_public returns DataPutPublicResult" do
    {:ok, result} =
      simulate_grpc_call(:ok, fn ->
        %Antd.DataPutPublicResult{address: "abc123"}
      end)

    assert result.address == "abc123"
  end

  test "data_get_public returns binary data" do
    {:ok, data} = simulate_grpc_call(:ok, fn -> "hello" end)
    assert data == "hello"
  end

  # ---------------------------------------------------------------------------
  # Data Private
  # ---------------------------------------------------------------------------

  test "data_put returns DataPutResult" do
    {:ok, result} =
      simulate_grpc_call(:ok, fn ->
        %Antd.DataPutResult{data_map: "dm123"}
      end)

    assert result.data_map == "dm123"
  end

  test "data_get returns binary data" do
    {:ok, data} = simulate_grpc_call(:ok, fn -> "secret" end)
    assert data == "secret"
  end

  # ---------------------------------------------------------------------------
  # Data Streaming
  #
  # Mirrors map_chunk_stream/1 in GrpcClient: a server-stream of
  # {:ok, DataChunk} (or raw struct) mapped to its binary payloads. Uses plain
  # maps in place of the proto struct so the test needs no generated module.
  # ---------------------------------------------------------------------------

  defp simulate_chunk_stream(replies) do
    Stream.map(replies, fn
      {:ok, %{data: data}} -> data
      %{data: data} -> data
    end)
  end

  test "data_stream yields concatenable binary chunks" do
    # Two chunks so chunk-boundary handling is exercised, not a single message.
    chunks = simulate_chunk_stream([{:ok, %{data: "sec"}}, {:ok, %{data: "ret"}}]) |> Enum.to_list()
    assert chunks == ["sec", "ret"]
    assert Enum.join(chunks) == "secret"
  end

  test "data_stream_public yields concatenable binary chunks" do
    assert simulate_chunk_stream([{:ok, %{data: "hel"}}, {:ok, %{data: "lo"}}]) |> Enum.join() ==
             "hello"
  end

  # Mirrors map_frame_stream/1 + frame_of/1, but over the REAL generated
  # `Antd.V1.DataChunk` oneof struct so the `kind: {:data | :progress, _}`
  # representation is genuinely exercised.
  defp simulate_frame_stream(chunks) do
    Stream.map(chunks, fn
      %Antd.V1.DataChunk{kind: {:progress, p}} ->
        %Antd.DownloadFrame{
          progress: %Antd.DownloadProgress{phase: p.phase, fetched: p.fetched, total: p.total}
        }

      %Antd.V1.DataChunk{kind: {:data, data}} ->
        %Antd.DownloadFrame{data: data}
    end)
  end

  test "data_stream_with_progress maps the oneof to data + progress frames" do
    frames =
      simulate_frame_stream([
        %Antd.V1.DataChunk{
          kind: {:progress, %Antd.V1.DownloadProgress{phase: "fetching", fetched: 1, total: 2}}
        },
        %Antd.V1.DataChunk{kind: {:data, "sec"}},
        %Antd.V1.DataChunk{kind: {:data, "ret"}}
      ])
      |> Enum.to_list()

    data = frames |> Enum.reject(&Antd.DownloadFrame.progress?/1) |> Enum.map_join(& &1.data)
    progress = Enum.filter(frames, &Antd.DownloadFrame.progress?/1)

    assert data == "secret"
    assert [%Antd.DownloadFrame{progress: %Antd.DownloadProgress{phase: "fetching", fetched: 1, total: 2}}] =
             progress
  end

  # ---------------------------------------------------------------------------
  # PaymentMode serialization (atom -> wire string)
  # ---------------------------------------------------------------------------

  test "PaymentMode.to_wire/1 serializes atoms to wire strings" do
    assert Antd.PaymentMode.to_wire(:auto) == "auto"
    assert Antd.PaymentMode.to_wire(:merkle) == "merkle"
    assert Antd.PaymentMode.to_wire(:single) == "single"
    assert Antd.PaymentMode.to_wire(nil) == "auto"
  end

  # ---------------------------------------------------------------------------
  # Data Cost
  # ---------------------------------------------------------------------------

  test "data_cost returns cost string" do
    {:ok, cost} = simulate_grpc_call(:ok, fn -> "50" end)
    assert cost == "50"
  end

  # ---------------------------------------------------------------------------
  # Chunks
  # ---------------------------------------------------------------------------

  test "chunk_put returns PutResult" do
    {:ok, result} =
      simulate_grpc_call(:ok, fn ->
        %Antd.PutResult{cost: "10", address: "chunk1"}
      end)

    assert result.cost == "10"
    assert result.address == "chunk1"
  end

  test "chunk_get returns binary data" do
    {:ok, data} = simulate_grpc_call(:ok, fn -> "chunkdata" end)
    assert data == "chunkdata"
  end

  # ---------------------------------------------------------------------------
  # Files & Directories
  # ---------------------------------------------------------------------------

  test "file_put_public returns FilePutPublicResult" do
    {:ok, result} =
      simulate_grpc_call(:ok, fn ->
        %Antd.FilePutPublicResult{
          address: "file1",
          storage_cost_atto: "1000",
          gas_cost_wei: "42",
          chunks_stored: 3,
          payment_mode_used: "auto"
        }
      end)

    assert result.address == "file1"
    assert result.storage_cost_atto == "1000"
    assert result.gas_cost_wei == "42"
    assert result.chunks_stored == 3
    assert result.payment_mode_used == "auto"
  end

  test "file_put returns FilePutResult" do
    {:ok, result} =
      simulate_grpc_call(:ok, fn ->
        %Antd.FilePutResult{
          data_map: "dm_file_1",
          storage_cost_atto: "500",
          gas_cost_wei: "21",
          chunks_stored: 2,
          payment_mode_used: "single"
        }
      end)

    assert result.data_map == "dm_file_1"
    assert result.storage_cost_atto == "500"
    assert result.gas_cost_wei == "21"
    assert result.chunks_stored == 2
    assert result.payment_mode_used == "single"
  end

  test "file_get_public returns :ok" do
    {:ok, result} = simulate_grpc_call(:ok, fn -> :ok end)
    assert result == :ok
  end

  test "file_get returns :ok" do
    {:ok, result} = simulate_grpc_call(:ok, fn -> :ok end)
    assert result == :ok
  end

  test "file_cost returns cost string" do
    {:ok, cost} = simulate_grpc_call(:ok, fn -> "1000" end)
    assert cost == "1000"
  end

  # ---------------------------------------------------------------------------
  # Error mapping: GRPC.RPCError -> Antd error structs
  # ---------------------------------------------------------------------------

  test "INVALID_ARGUMENT -> BadRequestError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 3, message: "bad arg"})
    assert %Antd.BadRequestError{} = err
    assert err.status_code == 400
    assert err.message == "bad arg"
  end

  test "NOT_FOUND -> NotFoundError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 5, message: "not found"})
    assert %Antd.NotFoundError{} = err
    assert err.status_code == 404
  end

  test "ALREADY_EXISTS -> AlreadyExistsError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 6, message: "exists"})
    assert %Antd.AlreadyExistsError{} = err
    assert err.status_code == 409
  end

  test "RESOURCE_EXHAUSTED -> TooLargeError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 8, message: "too big"})
    assert %Antd.TooLargeError{} = err
    assert err.status_code == 413
  end

  test "INTERNAL -> InternalError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 13, message: "crash"})
    assert %Antd.InternalError{} = err
    assert err.status_code == 500
  end

  test "UNAVAILABLE -> NetworkError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 14, message: "down"})
    assert %Antd.NetworkError{} = err
    assert err.status_code == 502
  end

  test "FAILED_PRECONDITION -> PaymentError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 9, message: "no funds"})
    assert %Antd.PaymentError{} = err
    assert err.status_code == 402
  end

  test "unknown gRPC code -> AntdError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 15, message: "data loss"})
    assert %Antd.AntdError{} = err
    assert err.status_code == 15
    assert err.message == "data loss"
  end

  # ---------------------------------------------------------------------------
  # Bang variants raise on error
  # ---------------------------------------------------------------------------

  test "bang variant raises BadRequestError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 3, message: "bad"})
    assert_raise Antd.BadRequestError, fn -> raise err end
  end

  test "bang variant raises NotFoundError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 5, message: "gone"})
    assert_raise Antd.NotFoundError, fn -> raise err end
  end

  test "bang variant raises InternalError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 13, message: "boom"})
    assert_raise Antd.InternalError, fn -> raise err end
  end

  test "bang variant raises PaymentError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 9, message: "funds"})
    assert_raise Antd.PaymentError, fn -> raise err end
  end

  test "bang variant raises AlreadyExistsError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 6, message: "dup"})
    assert_raise Antd.AlreadyExistsError, fn -> raise err end
  end

  test "bang variant raises TooLargeError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 8, message: "huge"})
    assert_raise Antd.TooLargeError, fn -> raise err end
  end

  test "bang variant raises NetworkError" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 14, message: "net"})
    assert_raise Antd.NetworkError, fn -> raise err end
  end

  test "bang variant raises AntdError for unknown code" do
    {:error, err} = simulate_grpc_call(:error, %GRPC.RPCError{status: 99, message: "???"})
    assert_raise Antd.AntdError, fn -> raise err end
  end

  # ---------------------------------------------------------------------------
  # Model immutability sanity checks
  # ---------------------------------------------------------------------------

  test "PutResult struct fields are accessible" do
    result = %Antd.PutResult{cost: "100", address: "abc"}
    assert result.cost == "100"
    assert result.address == "abc"
  end
end

File: ant-sdk/antd-dart/test/grpc_client_test.dart
import 'dart:typed_data';

import 'package:antd/src/errors.dart';
import 'package:antd/src/models.dart';
import 'package:test/test.dart';

// ---------------------------------------------------------------------------
// Standalone fake gRPC client for testing.
//
// Does NOT import grpc_client.dart (which requires proto-generated stubs).
// Instead, defines a _FakeGrpcClient with the same method surface that
// returns canned responses or throws fake gRPC-like errors for error mapping
// tests.
// ---------------------------------------------------------------------------

/// Simulates a gRPC error with a status code and message.
class FakeGrpcError implements Exception {
  final int code;
  final String message;
  FakeGrpcError(this.code, this.message);
}

/// Maps a fake gRPC status code to the SDK error types (mirrors the real
/// GrpcAntdClient._handleError logic).
Exception mapGrpcError(FakeGrpcError e) {
  switch (e.code) {
    case 3: return BadRequestError(e.message);
    case 5: return NotFoundError(e.message);
    case 6: return AlreadyExistsError(e.message);
    case 8: return TooLargeError(e.message);
    case 9: return PaymentError(e.message);
    case 13: return InternalError(e.message);
    case 14: return NetworkError(e.message);
    default: return AntdError(e.code, e.message);
  }
}

/// Fake gRPC client returning canned responses for all methods.
class _FakeGrpcClient {
  final FakeGrpcError? errorToThrow;

  _FakeGrpcClient({this.errorToThrow});

  Future<T> _maybeThrow<T>(T value) async {
    if (errorToThrow != null) throw mapGrpcError(errorToThrow!);
    return value;
  }

  Future<HealthStatus> health() =>
      _maybeThrow(const HealthStatus(ok: true, network: 'local'));

  Future<DataPutPublicResult> dataPutPublic(Uint8List data,
          {PaymentMode paymentMode = PaymentMode.auto}) =>
      _maybeThrow(const DataPutPublicResult(address: 'abc123'));

  Future<Uint8List> dataGetPublic(String address) =>
      _maybeThrow(Uint8List.fromList([104, 101, 108, 108, 111])); // "hello"

  Future<DataPutResult> dataPut(Uint8List data,
          {PaymentMode paymentMode = PaymentMode.auto}) =>
      _maybeThrow(const DataPutResult(dataMap: 'dm123'));

  Future<Uint8List> dataGet(String dataMap) =>
      _maybeThrow(Uint8List.fromList([115, 101, 99, 114, 101, 116])); // "secret"

  // Server-streams the payload in two chunks so the caller's chunk-by-chunk
  // consumption is exercised, not just a single message.
  Stream<List<int>> dataStream(String dataMap) async* {
    if (errorToThrow != null) throw mapGrpcError(errorToThrow!);
    yield [115, 101, 99]; // "sec"
    yield [114, 101, 116]; // "ret"
  }

  Stream<List<int>> dataStreamPublic(String address) async* {
    if (errorToThrow != null) throw mapGrpcError(errorToThrow!);
    yield [104, 101, 108]; // "hel"
    yield [108, 111]; // "lo"
  }

  // Mirrors GrpcAntdClient.dataStreamWithProgress: the byte total arrives first
  // (from the x-content-length header, surfaced as a meta frame), then when
  // include_progress is set the daemon interleaves DownloadProgress frames with
  // the data frames. Emits meta, a progress frame, then the data in two chunks.
  Stream<DownloadFrame> dataStreamWithProgress(String dataMap) async* {
    if (errorToThrow != null) throw mapGrpcError(errorToThrow!);
    yield const DownloadFrame.meta(6); // "secret"
    yield const DownloadFrame.progress(
        DownloadProgress(phase: 'fetching', fetched: 1, total: 2));
    yield DownloadFrame.data([115, 101, 99]); // "sec"
    yield DownloadFrame.data([114, 101, 116]); // "ret"
  }

  Stream<DownloadFrame> dataStreamPublicWithProgress(String address) async* {
    if (errorToThrow != null) throw mapGrpcError(errorToThrow!);
    yield const DownloadFrame.meta(5); // "hello"
    yield const DownloadFrame.progress(
        DownloadProgress(phase: 'resolving_map', fetched: 0, total: 0));
    yield DownloadFrame.data([104, 101, 108]); // "hel"
    yield DownloadFrame.data([108, 111]); // "lo"
  }

  Future<String> dataCost(Uint8List data,
          {PaymentMode paymentMode = PaymentMode.auto}) =>
      _maybeThrow('50');

  Future<PutResult> chunkPut(Uint8List data) =>
      _maybeThrow(const PutResult(cost: '10', address: 'chunk1'));

  Future<Uint8List> chunkGet(String address) =>
      _maybeThrow(Uint8List.fromList([99, 104, 117, 110, 107])); // "chunk"

  Future<FilePutResult> filePut(String path,
          {PaymentMode paymentMode = PaymentMode.auto}) =>
      _maybeThrow(const FilePutResult(
        dataMap: 'private_dm',
        storageCostAtto: '500',
        gasCostWei: '21',
        chunksStored: 2,
        paymentModeUsed: 'single',
      ));

  Future<void> fileGet(String dataMap, String destPath) => _maybeThrow(null);

  Future<FilePutPublicResult> filePutPublic(String path,
          {PaymentMode paymentMode = PaymentMode.auto}) =>
      _maybeThrow(const FilePutPublicResult(
        address: 'file1',
        storageCostAtto: '1000',
        gasCostWei: '42',
        chunksStored: 3,
        paymentModeUsed: 'auto',
      ));

  Future<void> fileGetPublic(String address, String destPath) =>
      _maybeThrow(null);

  Future<String> fileCost(
    String path, {
    bool isPublic = true,
    PaymentMode paymentMode = PaymentMode.auto,
  }) =>
      _maybeThrow('1000');

  Future<void> close() async {}
}

_FakeGrpcClient errorClient(int grpcCode, String message) {
  return _FakeGrpcClient(
    errorToThrow: FakeGrpcError(grpcCode, message),
  );
}

void main() {
  // -------------------------------------------------------------------------
  // Happy-path tests
  // -------------------------------------------------------------------------

  group('Health', () {
    test('returns health status', () async {
      final client = _FakeGrpcClient();
      final h = await client.health();
      expect(h.ok, isTrue);
      expect(h.network, equals('local'));
    });
  });

  group('Data Public', () {
    test('put public data', () async {
      final client = _FakeGrpcClient();
      final result = await client.dataPutPublic(
        Uint8List.fromList([1, 2, 3]),
        paymentMode: PaymentMode.merkle,
      );
      expect(result.address, equals('abc123'));
    });

    test('get public data', () async {
      final client = _FakeGrpcClient();
      final data = await client.dataGetPublic('abc123');
      expect(String.fromCharCodes(data), equals('hello'));
    });
  });

  group('Data Private', () {
    test('put private data', () async {
      final client = _FakeGrpcClient();
      final result = await client.dataPut(Uint8List.fromList([1]));
      expect(result.dataMap, equals('dm123'));
    });

    test('get private data', () async {
      final client = _FakeGrpcClient();
      final data = await client.dataGet('dm123');
      expect(String.fromCharCodes(data), equals('secret'));
    });

    test('stream private data', () async {
      final client = _FakeGrpcClient();
      final chunks = await client.dataStream('dm123').toList();
      final bytes = chunks.expand((c) => c).toList();
      expect(String.fromCharCodes(bytes), equals('secret'));
    });

    test('stream public data', () async {
      final client = _FakeGrpcClient();
      final chunks = await client.dataStreamPublic('abc123').toList();
      final bytes = chunks.expand((c) => c).toList();
      expect(String.fromCharCodes(bytes), equals('hello'));
    });

    test('stream private data with progress surfaces meta then progress',
        () async {
      final client = _FakeGrpcClient();
      final frames = await client.dataStreamWithProgress('dm123').toList();

      // Byte total (x-content-length) is surfaced first, before any data.
      expect(frames.first.isMeta, isTrue);
      expect(frames.first.totalBytes, equals(6));

      // Then the progress frame.
      expect(frames[1].isProgress, isTrue);
      expect(frames[1].progress!.phase, equals('fetching'));
      expect(frames[1].progress!.fetched, equals(1));
      expect(frames[1].progress!.total, equals(2));

      final bytes = frames
          .where((f) => f.data != null)
          .expand((f) => f.data!)
          .toList();
      expect(String.fromCharCodes(bytes), equals('secret'));
    });

    test('stream public data with progress surfaces meta then progress',
        () async {
      final client = _FakeGrpcClient();
      final frames =
          await client.dataStreamPublicWithProgress('abc123').toList();

      expect(frames.first.isMeta, isTrue);
      expect(frames.first.totalBytes, equals(5));

      expect(frames[1].isProgress, isTrue);
      expect(frames[1].progress!.phase, equals('resolving_map'));

      final bytes = frames
          .where((f) => f.data != null)
          .expand((f) => f.data!)
          .toList();
      expect(String.fromCharCodes(bytes), equals('hello'));
    });
  });

  group('Cost', () {
    test('data cost', () async {
      final client = _FakeGrpcClient();
      final cost = await client.dataCost(Uint8List.fromList([1, 2, 3]));
      expect(cost, equals('50'));
    });

    test('file cost', () async {
      final client = _FakeGrpcClient();
      final cost = await client.fileCost('/tmp/test.txt', isPublic: true);
      expect(cost, equals('1000'));
    });
  });

  group('Chunks', () {
    test('put chunk', () async {
      final client = _FakeGrpcClient();
      final result = await client.chunkPut(Uint8List.fromList([1, 2]));
      expect(result.cost, equals('10'));
      expect(result.address, equals('chunk1'));
    });

    test('get chunk', () async {
      final client = _FakeGrpcClient();
      final data = await client.chunkGet('chunk1');
      expect(String.fromCharCodes(data), equals('chunk'));
    });
  });

  group('Files Public', () {
    test('put public file', () async {
      final client = _FakeGrpcClient();
      final result = await client.filePutPublic('/tmp/test.txt');
      expect(result.address, equals('file1'));
      expect(result.storageCostAtto, equals('1000'));
      expect(result.gasCostWei, equals('42'));
      expect(result.chunksStored, equals(3));
      expect(result.paymentModeUsed, equals('auto'));
    });

    test('get public file', () async {
      final client = _FakeGrpcClient();
      await client.fileGetPublic('file1', '/tmp/out.txt');
      // No exception means success.
    });
  });

  group('Files Private', () {
    test('put private file', () async {
      final client = _FakeGrpcClient();
      final result = await client.filePut('/tmp/test.txt');
      expect(result.dataMap, equals('private_dm'));
      expect(result.chunksStored, equals(2));
    });

    test('get private file', () async {
      final client = _FakeGrpcClient();
      await client.fileGet('dm123', '/tmp/out.txt');
    });
  });

  // -------------------------------------------------------------------------
  // Error mapping tests
  // -------------------------------------------------------------------------

  group('Error Mapping (gRPC -> AntdError)', () {
    test('INVALID_ARGUMENT -> BadRequestError', () async {
      final client = errorClient(3, 'bad arg');
      expect(
        () => client.health(),
        throwsA(isA<BadRequestError>().having((e) => e.message, 'message', 'bad arg')),
      );
    });

    test('NOT_FOUND -> NotFoundError', () async {
      final client = errorClient(5, 'not found');
      expect(
        () => client.health(),
        throwsA(isA<NotFoundError>().having((e) => e.statusCode, 'statusCode', 404)),
      );
    });

    test('ALREADY_EXISTS -> AlreadyExistsError', () async {
      final client = errorClient(6, 'exists');
      expect(
        () => client.health(),
        throwsA(isA<AlreadyExistsError>()),
      );
    });

    test('RESOURCE_EXHAUSTED -> TooLargeError', () async {
      final client = errorClient(8, 'too big');
      expect(
        () => client.health(),
        throwsA(isA<TooLargeError>()),
      );
    });

    test('INTERNAL -> InternalError', () async {
      final client = errorClient(13, 'crash');
      expect(
        () => client.health(),
        throwsA(isA<InternalError>()),
      );
    });

    test('UNAVAILABLE -> NetworkError', () async {
      final client = errorClient(14, 'down');
      expect(
        () => client.health(),
        throwsA(isA<NetworkError>()),
      );
    });

    test('FAILED_PRECONDITION -> PaymentError', () async {
      final client = errorClient(9, 'no funds');
      expect(
        () => client.health(),
        throwsA(isA<PaymentError>()),
      );
    });

    test('unknown gRPC code -> AntdError with code', () async {
      final client = errorClient(15, 'data loss');
      expect(
        () => client.health(),
        throwsA(isA<AntdError>().having(
            (e) => e.statusCode, 'statusCode', 15)),
      );
    });

    test('error propagates from dataPutPublic', () async {
      final client = errorClient(5, 'missing data');
      expect(
        () => client.dataPutPublic(Uint8List.fromList([1])),
        throwsA(isA<NotFoundError>()),
      );
    });

    test('error propagates from chunkGet', () async {
      final client = errorClient(13, 'boom');
      expect(
        () => client.chunkGet('addr'),
        throwsA(isA<InternalError>()),
      );
    });

    test('error propagates from filePutPublic', () async {
      final client = errorClient(8, 'too large');
      expect(
        () => client.filePutPublic('/tmp/big.bin'),
        throwsA(isA<TooLargeError>()),
      );
    });
  });
}
